### PR TITLE
Implement topology utils methods

### PR DESCRIPTION
**Pull request by guissalustiano**
As discussed [here](https://github.com/bgpkit/valley-free/pull/7#issuecomment-1894902313). This PR creates methods to analyze the graph.
This is the naive implementation based on the examples folder.
I didn't like this implementation because for this method you always need the run the `paths_graph` before, maybe we can create another type `ValleyFreeTopology` that has the guarantee that is a DAG, only been generated by the `paths_graph` and already having the source is inside of them, what do you think?
I also found a bug in the `paths_graph` implementation in the test for the other methods that I will solve here.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -441,4 +583,91 @@ mod test {
...
+    #[ignore]
+    fn test_path_graph_never_generate_ciclic() {
+        let topo = Topology::from_caida(get_caida_data()).unwrap();
+
+        topo.all_asns().into_par_iter().for_each(|asn| {
+            let topo = topo.paths_graph(asn);
+            assert!(!is_cyclic_directed(&topo.graph));
+        });
+    }
+
```
Comment (guissalustiano) [inline src/lib.rs:636]: It is a very heavy test, I ran this locally for 1hour and didn't find any cycle.
I'm not sure if it is working to keep this here, but maybe could be useful in the future

**Comment by guissalustiano** [comment]
This PR are fixing and add a feature, I break this in two different PR to become easier to review
#9 #10

**Comment by digizeph** [comment]
I see that you are still pushing new commits to the PRs. Let me know if they're ready for review. Thanks!

**Comment by guissalustiano** [comment]
Hey @digizeph 
It's ready now! Thanks

**Pull request closed without merging**